src/java.base/share/native/launcher/main.c

```diff
@@ -167,6 +167,49 @@ static void setup_sighandler() {
     }
 }
 
+static const char *last_pid_filename = "/proc/sys/kernel/ns_last_pid";
+
+static int get_last_pid() {
+    FILE *last_pid_file = fopen(last_pid_filename, "r");
+    if (!last_pid_file) {
+        perror("last_pid_file fopen");
+        return -1;
+    }
+    int last_pid;
+    if (0 >= fscanf(last_pid_file, "%d", &last_pid)) {
+        fclose(last_pid_file);
+        perror("last_pid_file fscanf");
+        last_pid = -1;
+    }
+    fclose(last_pid_file);
+    return last_pid;
+}
+
+static void set_last_pid(int pid) {
+    FILE *last_pid_file = fopen(last_pid_filename, "w");
+    if (!last_pid_file) {
+        perror("last_pid_file fopen");
+        return;
+    }
+    if (0 > fprintf(last_pid_file, "%d", pid)) {
```

Comment:
Honestly I'd prefer write() to fprintf, to eliminate any possibility of buffering to write partial value to the kernel-exposed file.

Reply:
It seems like I've just met a case when `fprintf` returns ok (i.e. the correct number of bytes written), but the file content isn't changed. `strace`'ing shows that underlying `write` call returns `EPERM`, which probably wasn't forward to a caller.